src/cobra/core/reaction.py

```diff
@@ -590,46 +589,65 @@ def genes(self) -> FrozenSet:
         """
         return frozenset(self._genes)
 
-    def _update_genes_from_gpr(self, new_gene_names: Optional[Set] = None) -> None:
+    def _update_genes_from_gpr(self) -> None:
         """Update genes of reation based on GPR.
 
-        Parameters
-        ----------
-        new_gene_names: set
+        If the reaction has a model, and new genes appear in the GPR, they will be
+        created as Gene() entities and added to the model. If the reaction doesn't have
+        a model, genes will be created without a model.
+
+        Genes that no longer appear in the GPR will be removed from the reaction, but
+        not the model. If you want to remove them expliclty, use model.remove_genes().
         """
-        if new_gene_names is None:
-            if self._gpr.body is not None:
-                new_gene_names = self._gpr.genes
-            else:
-                new_gene_names = set()
+        context = get_context(self)
+        if self._gpr.body is not None:
+            new_gene_names = self._gpr.genes
+        else:
+            new_gene_names = set()
         old_genes = self._genes.copy()
+        new_genes = set()
         if self._model is None:
             self._genes = {Gene(i) for i in new_gene_names}
         else:
             model_genes = self._model.genes
             self._genes = set()
             for g_id in new_gene_names:
-                if model_genes.has_id(g_id):
-                    self._genes.add(model_genes.get_by_id(g_id))
-                else:
+                if not model_genes.has_id(g_id):
                     new_gene = Gene(g_id)
                     new_gene._model = self._model
-                    self._genes.add(new_gene)
                     model_genes.append(new_gene)
+                    if context:
+                        # Remove the gene later
+                        context(
+                            partial(
+                                remove_genes,
+                                model=self._model,
+                                gene_list=[model_genes.get_by_id(g_id)],
+                                remove_reactions=False,
+                            )
+                        )
+                        context(partial(setattr, new_gene, "_model", None))
+                        # Maybe should be
+                        # context(partial(self._model.genes.__isub__, [new_gene]))
+                new_gene = model_genes.get_by_id(g_id)
+                self._genes.add(new_gene)
```

Comment:
I left it as is, because the below comment will do it, and I do need the new_genes set for comparison. In this case, it will try to add new genes to rxn._genes twice, but that doesn't matter since it is a set.

Reply:
Sorry, I could have deleted self._genes.add(new_gene) and left the other. If you'd like to delete that, please delete the context below it. Or I could do it.

Reply:
Not sure how it can be fixed, since I need self._genes in line 637, and I need new_genes. Maybe it can be revised somehow, but I'm not sure. The ways I tired just led to errors.